test/Generics/constrained_type_witnesses.swift

```diff
@@ -0,0 +1,75 @@
+// RUN: %target-typecheck-verify-swift
+
+protocol P {
+  associatedtype A
+  // expected-note@-1 3{{protocol requires nested type 'A'; do you want to add it?}}
+}
+
+struct S1<T> {}
+
+extension S1 where T : P {
+  typealias A = Int
+}
+
+// This is rejected because S1.A is not a suitable witness for P.A.
+extension S1 : P {}
+// expected-error@-1 {{type 'S1<T>' does not conform to protocol 'P'}}
+
+struct S2<T> {}
+
+extension S2 where T : P {
+  typealias A = Never
+}
+
+// Hack: This is OK to make SwiftUI work, which accidentally relies on the
+// incorrect behavior with a typealias whose underlying type is 'Never'
+// (so it didn't hit the compiler crash).
+extension S2 : P {}
+
+// Here we have a suitable witness
+struct S3<T> {}
+
+extension S3 where T == Int {
+  typealias A = Int
+}
+
+extension S3 : P where T == Int {}
```

Comment:
Could you change one of the cases or add a new one for a suitable witness where two extensions are constrained differently? For example, `T: Equatable` (type witness context) and `T == Int` (conformance context).

Reply:
I tried that, but we diagnose the second typealias as a re-declaration even if the requirements in the where clause were mutually exclusive. We might not want to change this behavior.

I also tried putting the typealias in a constrained protocol extension, but that hits a circularity via the GSB. I'll investigate this issue later.

Reply:
Sorry for the confusion, what I meant is ~
```swift 
protocol P { associatedtype A }

struct Foo<T> {}

extension Foo where T: Equatable {
  typealias A = ...
}
extension Foo : P where T == Int {} // OK
```

Reply:
Ah, thanks for the explanation. I’ll add this in a follow-up commit since I have a few more changes coming in this area.